fsw/src/crypto.c

```diff
@@ -2647,46 +2648,252 @@ static int32 Crypto_PDU(char* ingest,TC_t* tc_frame)
 
     return status;
 }
-
-int32 Crypto_TC_ApplySecurity(char** ingest, int* len_ingest)
+int32 Crypto_TC_ApplySecurity(const uint8* in_frame, const uint32 in_frame_length, \
+    uint8 **enc_frame_test, uint32 *enc_frame_len)
 {
     // Local Variables
     int32 status = OS_SUCCESS;
-    unsigned char* tc_ingest = *ingest;
 
     #ifdef DEBUG
         OS_printf(KYEL "\n----- Crypto_TC_ApplySecurity START -----\n" RESET);
     #endif
 
-    // TODO: This whole function!
-    //len_ingest = len_ingest;
-    //ingest[0] = ingest[0];
-
-    int security_header_bytes = 18;
-    int security_trailer_bytes = 16;
-    int tc_size = *len_ingest + security_header_bytes + security_trailer_bytes;
+    #ifdef DEBUG
+        OS_printf("%d TF Bytes received\n", in_frame_length);
+        OS_printf("DEBUG - ");
+        for(int i=0; i < in_frame_length; i++)
+        {
+            OS_printf("%02X", ((uint8 *)&*in_frame)[i]);
+        }
+        OS_printf("\nPrinted %d bytes\n", in_frame_length);
+    #endif
 
-    unsigned char * tempTC=NULL;
-    tempTC = (unsigned char *)malloc(tc_size * sizeof (unsigned char));
-    CFE_PSP_MemSet(tempTC, 0, tc_size);
 
-    int count = 0;
-    //Create Security Header
-    for (;count < security_header_bytes;count++){
-        tempTC[count]= 0x55; //put dummy filler bits in security header for now.
+    TC_FramePrimaryHeader_t temp_tc_header;
+    // Primary Header
+    temp_tc_header.tfvn   = ((uint8)in_frame[0] & 0xC0) >> 6;
+    temp_tc_header.bypass = ((uint8)in_frame[0] & 0x20) >> 5;
+    temp_tc_header.cc     = ((uint8)in_frame[0] & 0x10) >> 4;
+    temp_tc_header.spare  = ((uint8)in_frame[0] & 0x0C) >> 2;
+    temp_tc_header.scid   = ((uint8)in_frame[0] & 0x03) << 8;
+    temp_tc_header.scid   = temp_tc_header.scid | (uint8)in_frame[1];
+    temp_tc_header.vcid   = ((uint8)in_frame[2] & 0xFC) >> 2;
+    temp_tc_header.fl     = ((uint8)in_frame[2] & 0x03) << 8;
+    temp_tc_header.fl     = temp_tc_header.fl | (uint8)in_frame[3];
+    temp_tc_header.fsn	  = (uint8)in_frame[4];
+
+    // Check if command frame flag set
+    if (temp_tc_header.cc == 1)
+    {
+        /*
+        ** CCSDS 232.0-B-3
+        ** Section 6.3.1
+        ** "Type-C frames do not have the Security Header and Security Trailer."
+        */
+        #ifdef TC_DEBUG
+            OS_printf(KYEL "DEBUG - Received Control/Command frame - nothing to do.\n" RESET);
+        #endif
```

Comment:
This is somewhat of an interesting condition because the enc_frame never gets filled out in this case... We should probably set a non-zero error code so that the calling application can recognize this state.

After that change, instead of the if-else, I would consider changing the else to the ( status == OS_SUCCESS ) conditional that is used throughout CryptoLib.

Reply:
Agrred. Fixed in upcoming merge. [cutright_dev 501fffd]